src/scheduler/work_bucket.rs

```diff
@@ -1,147 +1,198 @@
     }
-}
 
-impl<VM: VMBinding> Eq for PrioritizedWork<VM> {}
+    #[inline(always)]
+    fn steal_batch_and_pop(
+        &self,
+        dest: &Worker<Box<dyn GCWork<VM>>>,
+    ) -> Steal<Box<dyn GCWork<VM>>> {
+        self.queue.steal_batch_and_pop(dest)
+    }
 
-impl<VM: VMBinding> Ord for PrioritizedWork<VM> {
-    fn cmp(&self, other: &Self) -> cmp::Ordering {
-        self.priority.cmp(&other.priority)
+    #[inline(always)]
+    fn push(&self, w: Box<dyn GCWork<VM>>) {
+        self.queue.push(w);
     }
-}
 
-impl<VM: VMBinding> PartialOrd for PrioritizedWork<VM> {
-    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
-        Some(self.cmp(other))
+    #[inline(always)]
+    fn push_all(&self, ws: Vec<Box<dyn GCWork<VM>>>) {
+        for w in ws {
+            self.queue.push(w);
+        }
     }
 }
 
 pub struct WorkBucket<VM: VMBinding> {
     active: AtomicBool,
-    /// A priority queue
-    queue: RwLock<BinaryHeap<PrioritizedWork<VM>>>,
+    queue: BucketQueue<VM>,
+    prioritized_queue: Option<BucketQueue<VM>>,
     monitor: Arc<(Mutex<()>, Condvar)>,
     can_open: Option<Box<dyn (Fn(&GCWorkScheduler<VM>) -> bool) + Send>>,
+    group: Arc<WorkerGroup<VM>>,
 }
 
 impl<VM: VMBinding> WorkBucket<VM> {
-    pub const DEFAULT_PRIORITY: usize = 1000;
-    pub fn new(active: bool, monitor: Arc<(Mutex<()>, Condvar)>) -> Self {
+    pub fn new(
+        active: bool,
+        monitor: Arc<(Mutex<()>, Condvar)>,
+        group: Arc<WorkerGroup<VM>>,
+    ) -> Self {
         Self {
             active: AtomicBool::new(active),
-            queue: Default::default(),
+            queue: BucketQueue::new(),
+            prioritized_queue: None,
             monitor,
             can_open: None,
+            group,
         }
     }
+
+    #[inline(always)]
     fn notify_one_worker(&self) {
-        let _guard = self.monitor.0.lock().unwrap();
-        self.monitor.1.notify_one()
+        // If the bucket is not activated, don't notify anyone.
+        if !self.is_activated() {
+            return;
+        }
+        // Notify one if there're any parked workers.
+        if self.group.parked_workers() > 0 {
+            let _guard = self.monitor.0.lock().unwrap();
+            self.monitor.1.notify_one()
+        }
     }
-    fn notify_all_workers(&self) {
-        let _guard = self.monitor.0.lock().unwrap();
-        self.monitor.1.notify_all()
+
+    #[inline(always)]
+    pub fn notify_all_workers(&self) {
+        // If the bucket is not activated, don't notify anyone.
+        if !self.is_activated() {
+            return;
+        }
+        // Notify all if there're any parked workers.
+        if self.group.parked_workers() > 0 {
+            let _guard = self.monitor.0.lock().unwrap();
+            self.monitor.1.notify_all()
+        }
     }
+
+    #[inline(always)]
     pub fn is_activated(&self) -> bool {
-        self.active.load(Ordering::SeqCst)
+        self.active.load(Ordering::Relaxed)
```

Comment:
I am not sure about this.  Buckets are only updated when all workers are parked, so it is sure that `bucket.activate()` happens after all the `inc` operation of `pending_workers`.  However, there is no clear message for the waken workers to know that "if I see `bucket.is_actived() == true`, I see the effect of the execution of work packets of all previous buckets".  In fact, all workers can wake up spuriously (see my other comments).

Reply:
> if I see bucket.is_actived() == true, I see the effect of the execution of work packets of all previous buckets

This sounds like a strong assumption. To ensure this we will need to put a fence at the start and end of each packet (?).

Reply:
We don't need to put explicit fences, but atomic memory operations are effectively fences.  If an "acquire" load sees the value of a "release" store, the load happens after the store.  That's enough to ensure all other loads after that "acquire" load to happen after all other stores before the "release" store.

If that's still to hard to reason about, we can use `SeqCst`: all `SeqCst` atomic memory operations performed by all threads form **a global total order.**  If every worker **increments a counter** (parked workers) after they have done all their work, and some worker **activate bucket** only after **seeing the counter maxed** (equal to num of workers), and after waking up, workers only pick up packets from a bucket after **seeing it activated**, then we have the guarantee I mentioned.  Those atomic variables (the park counter, and the activation flag) are used for synchronization.

Yes. Just use `SeqCst` if not sure, as it is the easiest to reason about.